controllers/eventhandlers/targetgrouppolicy.go

```diff
@@ -0,0 +1,158 @@
+package eventhandlers
+
+import (
+	"context"
+	"github.com/aws/aws-application-networking-k8s/pkg/apis/applicationnetworking/v1alpha1"
+	"github.com/aws/aws-application-networking-k8s/pkg/model/core"
+	"github.com/aws/aws-application-networking-k8s/pkg/utils/gwlog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	gateway_api_v1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gateway_api "sigs.k8s.io/gateway-api/apis/v1beta1"
+	mcs_api "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+)
+
+type targetGroupPolicyEventHandler struct {
+	log    gwlog.Logger
+	client client.Client
+}
+
+func NewTargetGroupPolicyEventHandler(log gwlog.Logger, client client.Client) *targetGroupPolicyEventHandler {
+	return &targetGroupPolicyEventHandler{log: log, client: client}
+}
+
+func (h *targetGroupPolicyEventHandler) getTargetRef(obj client.Object) *corev1.Service {
+	tgp := obj.(*v1alpha1.TargetGroupPolicy)
+	policyName := tgp.Namespace + "/" + tgp.Name
+
+	targetRef := tgp.Spec.TargetRef
+	if targetRef.Group != "" || targetRef.Kind != "Service" {
+		h.log.Infow("Detected non-Service TargetGroupPolicy attachment, skipping",
+			"policyName", policyName, "targetRef", targetRef)
+		return nil
+	}
+	namespace := tgp.Namespace
+	if targetRef.Namespace != nil && namespace != string(*targetRef.Namespace) {
+		h.log.Infow("Detected cross namespace TargetGroupPolicy attachment, skipping",
+			"policyName", policyName, "targetRef", targetRef)
+		return nil
+	}
+
+	svcName := types.NamespacedName{
+		Namespace: namespace,
+		Name:      string(targetRef.Name),
+	}
+	svc := &corev1.Service{}
+	if err := h.client.Get(context.TODO(), svcName, svc); err != nil {
+		if errors.IsNotFound(err) {
+			h.log.Debugw("TargetGroupPolicy is referring to non-existent service, skipping",
+				"policyName", policyName, "serviceName", svcName.String())
+		} else {
+			// Still gracefully skipping the event but errors other than NotFound are bad sign.
+			h.log.Warnw("Failed to query targetRef of TargetGroupPolicy",
+				"policyName", policyName, "serviceName", svcName.String(), "reason", err.Error())
+		}
+		return nil
+	}
+	h.log.Debugw("TargetGroupPolicy change on Service detected",
+		"policyName", policyName, "serviceName", svcName.String())
+
+	return svc
+}
+
+func (h *targetGroupPolicyEventHandler) MapToHTTPRoute(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+
+	svc := h.getTargetRef(obj)
+	if svc == nil {
+		return nil
```

Comment:
Is there an external retry mechanisms for cases where we've errored internally and returned ```nil```?

Reply:
I don't think it is possible from what k8s provides, this applies for any other event handlers we have. I think our best bet is to log properly and users will hopefully notice and recreate resources